src/components/pages/rules/flyouts/ruleEditor/ruleEditor.js

```diff
@@ -0,0 +1,336 @@
+import { IoTHubManagerService, TelemetryService } from 'services';
+import { toNewRuleRequestModel } from 'services/models';
+
+import './ruleEditor.css';
+
+const Section = Flyout.Section;
+const severityLevels = ['critical', 'warning', 'info'];
+const calculations = ['average', 'instant'];
+var deviceGroupOptions = [];
+var operatorOptions = [
+  { label: '>', value: 'GreaterThan' },
+  { label: '>=', value: 'GreaterThanOrEqual' },
+  { label: '<', value: 'LessThan' },
+  { label: '<=', value: 'LessThanOrEqual' },
+  { label: '=', value: 'Equals' }
+];
+// A counter for creating unique keys per new condition
+let conditionKey = 0;
+
+// Creates a state object for a condition
+const newCondition = () => ({
+  field: '',
+  calculation: '',
+  duration: '00:00:00',
+  operator: operatorOptions[0].value,
+  value: '',
+  key: conditionKey++ // Used by react to track the rendered elements
+});
+
+// A state object for a new rule
+const newRule = {
+  name: '',
+  description: '',
+  groupId: '',
+  conditions: [newCondition()], // Start with one condition
+  severity: severityLevels[0],
+  enabled: true
+}
+
+export class RuleEditor extends LinkedComponent {
+
+  constructor(props) {
+    super(props);
+    const { deviceGroups = [], rule } = props;
+    const formData = rule ? rule : newRule;
+    this.state = {
+      isPending: false,
+      error: undefined,
+      fieldOptions: [],
+      devicesAffected: 0,
+      formData
+    };
+    deviceGroupOptions = deviceGroups.map(this.toSelectOption);
+  }
+
+  componentWillUnmount() {
+    if (this.subscription) this.subscription.unsubscribe();
+  }
+
+  toSelectOption = ({ id, displayName }) => ({ label: displayName, value: id });
+
+  addCondition = () => this.conditionsLink.set([...this.conditionsLink.value, newCondition()]);
+
+  deleteCondition = (index) =>
+    (evt) => this.conditionsLink.set(this.conditionsLink.value.filter((_, idx) => index !== idx));
+
+  formIsValid() {
+    return [
+      this.ruleNameLink,
+      this.deviceGroupLink,
+      this.conditionsLink,
+    ].every(link => !link.error);
+  }
+
+  //TODO: still working on it as of 17/04/18
+  apply = (event) => {
+    event.preventDefault();
+    const { formData } = this.state;
+    console.log('TODO: Handle the form submission');
+    if (this.formIsValid()) {
+      this.setState({ isPending: true });
+      this.subscription = TelemetryService.createRule(toNewRuleRequestModel(formData))
+        .subscribe(
+          () => {
+            this.setState({ isPending: false });
+            this.props.onClose();
+          },
+          error => this.setState({ error, isPending: false })
+        );
+    }
+  }
+
+  onGroupIdChange = ({ target: { value: { value = {} } } }) => {
+    this.getDeviceCountAndFields(value);
+  }
+
+  getDeviceCountAndFields(groupId) {
+    this.props.deviceGroups.some(group => {
+      if (group.id === groupId) {
+        this.subscription = IoTHubManagerService.getDevices(group.conditions)
```

Comment:
Need to check if the subscription has anything in it before overriding it. If something is there it should be unsubscribed.

Reply:
FYI - not done this way in new device flyout. I'll do here.